server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeMetaWriter.java

```diff
@@ -0,0 +1,226 @@
+            // segment aggregated document count
+            totalSizeInBytes += Integer.BYTES;
+            // max leaf docs
+            totalSizeInBytes += Integer.BYTES;
+            // skip star node creation dimensions count
+            totalSizeInBytes += Integer.BYTES;
+            // skip star node creation dimensions field numbers
+            totalSizeInBytes += (long) starTreeField.getStarTreeConfig().getSkipStarNodeCreationInDims().size() * Integer.BYTES;
+            // data start file pointer
+            totalSizeInBytes += Long.BYTES;
+            // data length
+            totalSizeInBytes += Long.BYTES;
+
+            logger.debug("Star tree size in bytes : {}", totalSizeInBytes);
+        }
+
+        writeMetaHeader(metaOut, compositeFieldType, starTreeField.getName());
+        writeMeta(metaOut, writeState, metricAggregatorInfos, starTreeField, segmentAggregatedCount, dataFilePointer, dataFileLength);
+    }
+
+    /**
+     * Computes the byte size required to store the star-tree metric entry.
+     *
+     * @param metricAggregatorInfos the list of metric aggregator information
+     * @return the byte size required to store the metric-metric stat pairs
+     */
+    private static long computeMetricEntriesSizeInBytes(List<MetricAggregatorInfo> metricAggregatorInfos) {
+
+        long totalMetricEntriesSize = 0;
+
+        for (MetricAggregatorInfo metricAggregatorInfo : metricAggregatorInfos) {
+            totalMetricEntriesSize += metricAggregatorInfo.getField().getBytes(UTF_8).length;
+            totalMetricEntriesSize += metricAggregatorInfo.getMetricStat().getTypeName().getBytes(UTF_8).length;
+        }
+
+        return totalMetricEntriesSize;
+    }
+
+    /**
+     * Computes the byte size of the star-tree metadata header.
+     *
+     * @param compositeFieldType the composite field type of the star-tree field
+     * @param starTreeFieldName  the name of the star-tree field
+     * @return the byte size of the star-tree metadata header
+     */
+    private static int computeHeaderByteSize(CompositeMappedFieldType.CompositeFieldType compositeFieldType, String starTreeFieldName) {
+        // Magic marker (8), version (4), size of header (4)
+        int headerSizeInBytes = 16;
+
+        // For star-tree field name
+        headerSizeInBytes += starTreeFieldName.getBytes(UTF_8).length;
+
+        // For star tree field type
+        headerSizeInBytes += compositeFieldType.getName().getBytes(UTF_8).length;
+
+        return headerSizeInBytes;
+    }
+
+    /**
+     * Writes the star-tree metadata header.
+     *
+     * @param metaOut            the IndexOutput to write the header
+     * @param compositeFieldType the composite field type of the star-tree field
+     * @param starTreeFieldName  the name of the star-tree field
+     * @throws IOException if an I/O error occurs while writing the header
+     */
+    private static void writeMetaHeader(
```

Comment:
Should these headers be present in composite99writer before we build a field ? Say we are about to build a field , then we can write these headers there

So the control will be at composite index level rather than at each implementation .
Anyways apart from this we write star tree specific metadata.

With inheritance implemented this can go into composite index metadata easily.

Reply:
> this can go into composite index metadata easily.

yeah, I am not sure about this. The thing is, header stores the composite field type, based on which we will be creating a star tree meta (extending composite index meta) object. So I need to read this outside of Composite Index Meta, and then create the object based on type.

Reply:
> Shouldn't this be driven by implementation. We would be able to support different implementation easily (like V1, V2 etc) instead of bringing this to index level. wdyt?

This is file format change - so , any change in implementations , we need to ideally increment the composite index codec version as well right ?

So, v1, v2 of star tree will require v1, v2 of composite codec as well - so lets make this specific to composite format / codec.

Type - if we introduce new type, we need a new composite codec version, so both of these should be visible in composite codec.